### PR TITLE
New use case: Onboarding of CNFs to the platform

### PR DESCRIPTION
**Pull request by vukg**
Here is new use case that covers challenges with packaging, distribution and deployment of CNFs - what we colloquially call CNF "onboarding". I am looking forward to hear the feedback from the community. There is for sure improvement potential for this PR, but there is already enough material for comments.

Bests,
Vuk

### CODE REVIEW

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continous deployment realisations with that particular or similar CNFs as well as comprihensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```
Comment (xmulligan) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+The starting point of onboarding shall be extension of platfrom with any requiremetns and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.
```
Comment (xmulligan) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
The starting point of onboarding shall be extension of platform with any requiremetns and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.
```

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+## Challenges and limitations with Kube-native approach
+
+Most of the CNFs today, especially complex ones, are packaged in such way that only trained professional service teams of the CNF vendor can reasonably deploy them. The documentation and manuals that come with those CNFs are giving same impression. The onboarding today usually requires intensive interation between CNF vendor, CNF DevOps and CNPD teams and lots of trial/error tuning and manual adaptations on both platform and CNF side.
+
+The CNFs are also regularly packaged to be suitable for opinionated installers that are typically part of vendor's own platform offering or particular platform blueprint. The onboarding on CSP's own platform in such cases frequently involves sort of reverse engineering of CNF vendor's installer to come to basic cloud native artefacts and derive requirements from installer's code with the help of vendor's engineers.
+
+Finally, the CNFs are frequently delivered as Helm charts. However, these Helm charts often do not cover all dependencies that enables reconciling the deployments in propper way. They also install CRDs which are not managed well by Helm.
```
Comment (xmulligan) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
Finally, the CNFs are frequently delivered as Helm charts. However, these Helm charts often do not cover all dependencies that enables reconciling the deployments in proper way. They also install CRDs which are not managed well by Helm.
```

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md (hunk 1 above)
Comment (electrocucaracha) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continous deployment realisations with that particular or similar CNFs as well as comprihensive deployment troubleshooting guide. As already mentioned in [UC0001](0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md), CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+This situation is in most cases not blocking the onboarding, but is making it laborious and not repeatable.
+
+### Established practices to overcome challenges and limitations
+
+The practices already exist outside of CNF world, but are not very well followed by CNF vendors. Some of the practices are here:
+* https://helm.sh/docs/chart_best_practices/
+* https://www.gitops.tech/
+* https://learnk8s.io/production-best-practices
```
Comment (electrocucaracha) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
- [The Chart Best Practices Guide](https://helm.sh/docs/chart_best_practices/)
- [GitOps](https://www.gitops.tech/)
- [Kubernetes production best practices](https://learnk8s.io/production-best-practices)
```

**Review by electrocucaracha — COMMENTED** [review]
Linting suggestions

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+In CNF case the assumption is that the platform is already there, that it is in its core Kubernetes based without special forks as well as that different components needed by CNF could be installed via Kubernetes API.
+
+This creates the situation in which CNF and CSP platform, both of which have their requirements and pre-conditions, meet each other as such for the first time. The process of getting the CNF deployed on the platform in aforementioned scenario is colloquially called "onboarding". In this process CNF vendor, CNF DevOps and CNPD work together to reach common gloal of having production grade deployment of CNF supported with appropriate SLA which factors-in the fact that CNF runs on CSP platform.  
+
+## Expected behaviour
+
+CNF vendor expects that CSP platform and environment (e.g. Network) are fulfiling certain requirements and preconditions for CNF to be able to achieve production grade deployment. These requirements are different from CNF to CNF.
```
Comment (CsatariGergely) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]: How about Anuket RA2. It's main target is to "standardize" the "certain requirements and preconditions".

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
+
+The starting point of onboarding shall be extension of platform with any requiremetns and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.
```
Comment (CsatariGergely) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
The starting point of onboarding shall be extension of platform with any requirements and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.
```

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+## Challenges and limitations with Kube-native approach
+
+Most of the CNFs today, especially complex ones, are packaged in such way that only trained professional service teams of the CNF vendor can reasonably deploy them. The documentation and manuals that come with those CNFs are giving same impression. The onboarding today usually requires intensive interation between CNF vendor, CNF DevOps and CNPD teams and lots of trial/error tuning and manual adaptations on both platform and CNF side.
```
Comment (CsatariGergely) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md:51]: Shouldn't be the DevOps team a cross organisational team?

**Review by CsatariGergely — CHANGES_REQUESTED** [review]

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```
Comment (xmulligan) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continuous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md (hunk 7 above)
Comment (vukg) [reply use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md:51]: Hi Gergely,
When CNF DevOps and Cloud-native Platform DevOps work together it is cross-organisational. It is also advisable that CNF DevOps team has also members who possess CKA-like knowledge and who can deal with deployments, but it depends on the size of the organisation.

Does that make sense?
-Vuk

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md (hunk 5 above)
Comment (vukg) [reply use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]: Anuket to my knowledge is perscriptive when it comes to selection of components and their versions. It does not provide generalized requirements. However, number of requirements could be derived from it.

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md (hunk 1 above)
Comment (vukg) [reply use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]: for some reason I can not commit your suggestion @electrocucaracha , but it is fine by me. Can someone e.g. @taylor committ it?

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+The CNFs are also regularly packaged to be suitable for opinionated installers that are typically part of vendor's own platform offering or particular platform blueprint. The onboarding on CSP's own platform in such cases frequently involves sort of reverse engineering of CNF vendor's installer to come to basic cloud native artefacts and derive requirements from installer's code with the help of vendor's engineers.
+
+Finally, the CNFs are frequently delivered as Helm charts. However, these Helm charts often do not cover all dependencies that enables reconciling the deployments in proper way. They also install CRDs which are not managed well by Helm.
+
+This situation is in most cases not blocking the onboarding, but is making it laborious and not repeatable.
+
+### Established practices to overcome challenges and limitations
+
+The practices already exist outside of CNF world, but are not very well followed by CNF vendors. Some of the practices are here:
```
Comment (xmulligan) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md:61]:
```suggestion
The practices already exist outside of CNF world, but are not very well followed by CNF vendors. Some of the practices are here:

```

Comment (xmulligan) [reply use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md:61]: for linting

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md (hunk 5 above)
Comment (iawells) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
CNF vendor expects that CSP platform and environment (e.g. Network) are fulfilling certain requirements and preconditions for CNF to be able to achieve production grade deployment. These requirements are different from CNF to CNF.
```

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md (hunk 6 above)
Comment (iawells) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md]:
```suggestion
The starting point of onboarding shall be extension of platform with any requirements and dependencies that are coming from CNF. When all conditions on the platform level are met, the deployment of CNF shall proceed.
```

**Review by jeffsaelens — APPROVED** [review]
This is a good starting point. I'll add the user story as an additional PR in the near future.

**Review by fkautz — APPROVED** [review]

**Review by taylor — APPROVED** [review]
lgtm

**Review by Smitholi67 — APPROVED** [review]

**Review by electrocucaracha — APPROVED** [review]

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+- [The Chart Best Practices Guide](https://helm.sh/docs/chart_best_practices/)
+- [GitOps](https://www.gitops.tech/)
+- [Kubernetes production best practices](https://learnk8s.io/production-best-practices)
+
+### What needs to be done differently in order to overcome challenges and limitations
+
+CNF vendors need to start preparing CNFs for shipping as regular cloud software that is well docummented and that can be deployed and configured by CNF DevOps team. Following established best practices for cloud native software distribution and deployment would improve the situation significantly.
```
Comment (tliron) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md:69]: Spelling mistake: "docummented"

File: use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+CNF DevOps team expects to deploy and configure the CNF application on their own, preferably directly via automation pipelines and in interaction and cooperation with CNPD and CSP Network teams. For that they expect to receive well documented installation procedures, references for typical continuous deployment realisations with that particular or similar CNFs as well as comprehensive deployment troubleshooting guide. As already mentioned in (UC0001)[0001-UC-lifecycle-of-infrastructure-where-CNF-is-running.md], CNF DevOps team also expects to get suite of tests from CNF vendor. Tests could then be deployed in continuous testing pipeline to automatically verify that CNF is running properly (production grade) on the platform and that the deployment qualifies for SLA.
```
Comment (tliron) [inline use-case/0004-UC-onboarding-of-CNF-to-the-CSP-platform.md:45]: Wrong markdown for UC001 link (first brackets, then parentheses)

**Pull request merged**